File: 189. Rotate Array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        My solution, O(n) time O(k) space 
        """
        if not nums:
            return
        k = k % len(nums)
        temp = nums[len(nums)-k:]
        nums[k:] = nums[:len(nums)-k]
        nums[:k] = temp

class Solution:
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        Using Cyclic Replacements O(n) time O(1) space
        """
        k = k % len(nums)
        count = 0
        start = 0
        while count < len(nums):
            current = start 
            prev = nums[start] #store the value in the position
            
            while True:
                next = (current + k) % len(nums) #
                temp = nums[next] #store it temporaly 
                nums[next] = prev #overide the next 
                prev = temp #reset prev
                current = next #reset current
                count += 1

                if start == current:
                    break 
            
            start += 1

class Solution:
    def rotate(self, nums, k) -> None:
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        First reverse all, then reverse first k, then reverse others
        O(n) time O(1) space 
        """
        k %= len(nums)
        self.reverse(nums,0,len(nums)-1)
        self.reverse(nums,0, k-1)
        self.reverse(nums,k, len(nums)-1)

    def reverse(self, nums, start, end) -> None:
        """
        :type nums: List[int]
        :type start: int
        :type end: int
        :rtype: None
        """
        while start < end: #
            temp = nums[start]
            nums[start] = nums[end]
            nums[end] = temp 
            start += 1
            end -= 1

File: 189. Rotate Array.java
/**
 * O(n) time O(1) space, rotate to final position
 */
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        int count = 0;
        for (int start=0; count<n; start++) {
            int curr = start;
            int prev = nums[start];
            while (true) {
                int next = (curr + k) % n;
                int temp = nums[next];
                nums[next] = prev;
                prev = temp;
                curr = next;
                count++;
                if (curr == start) {
                    break;
                }
            }
        }
    }
}

/**
 * reverse the whole array, and then reverse first part and second part
 * O(n) time O(1) space
 */
class Solution {
    public void rotate(int[] nums, int k) {
        reverse(nums, 0, nums.length-1);
        reverse(nums, 0, k % nums.length-1);
        reverse(nums, k % nums.length, nums.length-1);
    }
    private void reverse(int[] nums, int st, int ed) {
        while (st < ed) {
            int temp = nums[st];
            nums[st] = nums[ed];
            nums[ed] = temp;
            st++; ed--;
        }
    }
}